File: topics/4-principal-component-analysis/Wine.m
%% Data prep

clearvars
clc

X = csvread('Wine.csv');
class = X(:,1); X = X(:,2:end);
varNames = {'Alcohol','Malic acid','Ash','Alcalinity of ash','Magnesium', ...
            'Total phenols','Flavanoids','Nonflavanoid phenols', ...
            'Proanthocyanins','Color intensity','Hue', ...
            'OD280/OD315 of diluted wines','Proline'};

%% Scale data

N = size(X, 1);
D = size(X, 2);
Xscaled = X - repmat(mean(X), N, 1);
Xscaled = Xscaled./repmat(std(X), N, 1);

%% PCA

S = cov(Xscaled);
[U, lambda] = eig(S);

U = sortrows([diag(lambda) U'], -1);
lambda = diag(U(:,1));
U = U(:,2:end)';

Xpca = Xscaled*U;

%% Plot

scatter(-Xpca(:,1), Xpca(:,2), 'filled')

File: data/python/wine.py
from pathlib import Path
import numpy as np
import pandas
from pandas import plotting
import matplotlib.pyplot as plt

data_file = Path(__file__).parent / ".." / "raw" / "wine" / "Wine.csv"
data = pandas.read_csv(
    data_file,
    names=(
        "Alcohol",
        "Malic acid",
        "Ash",
        "Alcalinity of ash",
        "Magnesium",
        "Total phenols",
        "Flavanoids",
        "Nonflavanoid phenols",
        "Proanthocyanins",
        "Color intensity",
        "Hue",
        "OD280/OD315 of diluted wines",
        "Proline",
    ),
)


def main():
    plt.rcParams["figure.figsize"] = (
        np.array(plt.rcParams["figure.figsize"]) * len(data.columns)
    ).tolist()

    plotting.scatter_matrix(data)
    plt.show()


if __name__ == "__main__":
    main()
